### DISCUSSION TITLE
Unable to add Contact Geometry: ContactMesh obj file

### DISCUSSION THREAD

**Discussion by bfugger** [ComputationalBiomechanicsLab/opensim-creator#1059] [Q&A] [answered]
Hi there, I am so far unable to add a contactmesh to my project. I typed the path to the obj file and I ensured that it is watertight and closed and i selected the proper frame. When I click "add" I don't have any contact geometry added. There are no error or info lines in the log. Any help is greatly appreciated

**Comment by adamkewley** [comment, marked as answer]
Hi @bfugger

I confirmed the bug on my side. From what I can see:

1. After adding a `ContactMesh` to a model and choosing an STL or OBJ file
2. The component is added to the model fine (as you found)
3. And the resulting `osim` will load sucessfully in both OpenSim Creator and OpenSim GUI with no error message
4. But the mesh will not be shown in either UI

---

The bug appears to be from OpenSim:

- `ContactMesh` sourcecode in OpenSim is here:
    - https://github.com/opensim-org/opensim-core/blob/4.4/OpenSim/Simulation/Model/ContactMesh.h
- Contains two mesh pointers: one for contact and one for decoration:
    - https://github.com/opensim-org/opensim-core/blob/4.4/OpenSim/Simulation/Model/ContactMesh.h#L114
- The decorative one is only populated via either:
  - The full-fat constructor (won't be used by either OpenSim Creator or OpenSimGUI): https://github.com/opensim-org/opensim-core/blob/4.4/OpenSim/Simulation/Model/ContactMesh.cpp#L55
  - `loadMesh`, which might be called when adding the contact mesh into the physics system: https://github.com/opensim-org/opensim-core/blob/4.4/OpenSim/Simulation/Model/ContactMesh.cpp#L96
- `loadMesh` is called by `ContactMesh::createSimTKContactGeometry`, but `createSimTKContactGeometry` is only called by forces that use contact geometries
    - e.g. here: https://github.com/opensim-org/opensim-core/blob/4.4/OpenSim/Simulation/Model/HuntCrossleyForce.cpp#L87
- So the only way to visualize a `ContactMesh` in the 3D scene is to add the `ContactMesh` to a force, because the side-effect of doing that is that the decorative geometry will also be loaded

So, @bfugger, the way I was able to move forward on this was to add a `ContactMesh` to the model (works fine) and then add a `HuntCrossleyForce` to the model and set one of its `geometry` properties to the name of the `ContactMesh` (defaults to `ContactMesh`).

However, I couldn't move forward from there because SimTK seems to prescribe that any mesh file that is provided *must* contain triangles with edges that share exactly two faces and I couldn't find an appropriate mesh during testing.

Your bug (not being able to see the contact geometry until a force is added) isn't in OpenSim Creator. It will need to be upstreamed to OpenSim. The potential later "bug" (of having to provide a perfect mesh file - not necessarily a bug but might be very annoying) appears to originate from simbody, which is another upstream library that OpenSim uses:

- https://simtk.org/api_docs/molmodel/api_docs22/Simbody/html/classSimTK_1_1ContactGeometry_1_1TriangleMesh.html#_details
- Error source: https://github.com/simbody/simbody/blob/master/SimTKmath/Geometry/src/ContactGeometry_TriangleMesh.cpp#L464

**Comment by bfugger** [comment]
Thank you so much for the detailed write up. I also think that I may be attempting to use a noncompliant mesh. Some further digging on the forum and the links you provided show my mesh also needs to be onesided? and with proper vertices directions. So i am working on cleaning that up. In the meantime, I will try the approach you described as I need to assign forces anyways. I am still getting the hang of opensim so I appreciate all your input.

**Comment by adamkewley** [comment]
Moving this to a discussion because the core issue is related to how OpenSim works, rather than something that OpenSim Creator can solve in-tree.
